Mathlib/Combinatorics/SimpleGraph/Bipartite.lean

```diff
@@ -0,0 +1,227 @@
+/-
+Copyright (c) 2025 Mitchell Horner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mitchell Horner
+-/
+import Mathlib.Combinatorics.Enumerative.DoubleCounting
+import Mathlib.Combinatorics.SimpleGraph.DegreeSum
+
+/-!
+# Bipartite graphs
+
+This file proves results about bipartite simple graphs, including several double-counting arguments.
+
+## Main definitions
+
+* `SimpleGraph.IsBipartite G s t` is the condition that a simple graph `G` is bipartite in sets
+  `s`, `t`, that is, `s` and `t` are disjoint and vertices `v`, `w` being adjacent in `G` implies
+  that `v ∈ s` and `w ∈ t`, or `v ∈ s` and `w ∈ t`.
+
+  Note that in this implementation, if `G.IsBipartite s t`, `s ∪ t` need not cover the vertices of
+  `G`, instead `s ∪ t` is only required to cover the *support* of `G`, that is, the vertices that
+  form edges in `G`. This definition is equivalent to the expected definition. If `s` and `t` do
+  not cover all the vertices, one recovers a covering of all the vertices by unioning the missing
+  vertices `(s ∪ t)ᶜ` to either `s` or `t`.
+
+* `SimpleGraph.isBipartite_sum_degrees_eq` is the proof that if `G.IsBipartite s t`, then the sum
+  of the degrees of the vertices in `s` is equal to the sum of the degrees of the vertices in `t`.
+
+* `SimpleGraph.isBipartite_sum_degrees_eq_card_edges` is the proof that if `G.IsBipartite s t`,
+  then sum of the degrees of the vertices in `s` is equal to the number of edges in `G`.
+-/
+
+
+open BigOperators Finset Fintype
+
+namespace SimpleGraph
+
+variable {V : Type*} {v w : V} {G : SimpleGraph V} {s t : Set V}
+
+section IsBipartite
+
+/-- `G` is bipartite in sets `s` and `t` iff `s` and `t` are disjoint and if vertices `v` and `w`
+are adjacent in `G` then `v ∈ s` and `w ∈ t`, or `v ∈ t` and `w ∈ s`. -/
+structure IsBipartite (G : SimpleGraph V) (s t : Set V) : Prop where
```

Comment:
I'd call this `IsBipartiteWith`, since it also says what the two parts are. (Then `G.IsBipartite could be the fact that the graph is 2-colourable, eg)